src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java

```diff
@@ -85,77 +92,115 @@ public LocalActivityMarkerData build() {
+      this.replayTimeMillis = replayTimeMillis;
+      this.errReason = errReason;
+      this.attempt = attempt;
+      this.backoff = backoff;
+      this.isCancelled = isCancelled;
+    }
+  }
+
+  private final LocalActivityMarkerHeader headers;
   private final byte[] result;
-  private final long replayTimeMillis;
-  private final int attempt;
-  private final Duration backoff;
-  private final boolean isCancelled;
 
   private LocalActivityMarkerData(
       String activityId,
       String activityType,
       long replayTimeMillis,
       byte[] result,
       String errReason,
-      byte[] errJson,
       int attempt,
       Duration backoff,
       boolean isCancelled) {
-    this.activityId = activityId;
-    this.activityType = activityType;
-    this.replayTimeMillis = replayTimeMillis;
+    this.headers =
+        new LocalActivityMarkerHeader(
+            activityId, activityType, replayTimeMillis, errReason, attempt, backoff, isCancelled);
     this.result = result;
-    this.errReason = errReason;
-    this.errJson = errJson;
-    this.attempt = attempt;
-    this.backoff = backoff;
-    this.isCancelled = isCancelled;
+  }
+
+  private LocalActivityMarkerData(LocalActivityMarkerHeader headers, byte[] result) {
+    this.headers = headers;
+    this.result = result;
+    if (headers == null) {
+      System.out.println("test");
+    }
   }
 
   public String getActivityId() {
-    return activityId;
+    return headers.activityId;
   }
 
   public String getActivityType() {
-    return activityType;
+    return headers.activityType;
   }
 
   public String getErrReason() {
-    return errReason;
+    return headers.errReason;
   }
 
   public byte[] getErrJson() {
-    return errJson;
+    return Strings.isNullOrEmpty(headers.errReason) ? null : result;
   }
 
   public byte[] getResult() {
     return result;
   }
 
   public long getReplayTimeMillis() {
-    return replayTimeMillis;
+    return headers.replayTimeMillis;
   }
 
   public int getAttempt() {
-    return attempt;
+    return headers.attempt;
   }
 
   public Duration getBackoff() {
-    return backoff;
+    return headers.backoff;
   }
 
   public boolean getIsCancelled() {
-    return isCancelled;
+    return headers.isCancelled;
+  }
+
+  public Header getHeader() {
+    byte[] headerData = JsonDataConverter.getInstance().toData(headers);
+    Header header = new Header();
+    header.setFields(ImmutableMap.of(LOCAL_ACTIVITY_HEADER_KEY, ByteBuffer.wrap(headerData)));
+    return header;
+  }
+
+  public static LocalActivityMarkerData fromEventAttributes(
+      MarkerRecordedEventAttributes attributes) {
+    ByteBuffer byteBuffer = attributes.getHeader().getFields().get(LOCAL_ACTIVITY_HEADER_KEY);
+    byte[] bytes = org.apache.thrift.TBaseHelper.byteBufferToByteArray(byteBuffer);
+    LocalActivityMarkerHeader header =
+        JsonDataConverter.getInstance()
+            .fromData(bytes, LocalActivityMarkerHeader.class, LocalActivityMarkerHeader.class);
```

Comment:
We shouldn't hardcode the converter used to serialize/deserialize the header. The converter should be passed from the higher level code, probably as a parameter to fromEventAttributes and getHeader.

Reply:
Done